packages/core/src/actions/smartAccount/estimateUserOperationGas.ts

```diff
@@ -0,0 +1,94 @@
+import type { Chain, Client, Transport } from "viem";
+import {
+  type GetEntryPointFromAccount,
+  type SmartContractAccount,
+} from "../../account/smartContractAccount.js";
+import { isBaseSmartAccountClient } from "../../client/isSmartAccountClient.js";
+import { AccountNotFoundError } from "../../errors/account.js";
+import { IncompatibleClientError } from "../../errors/client.js";
+import type {
+  UserOperationEstimateGasResponse,
+  UserOperationStruct,
+} from "../../types.js";
+import {
+  conditionalReturn,
+  deepHexlify,
+  resolveProperties,
+  type Deferrable,
+} from "../../utils/index.js";
+import type {
+  SendUserOperationParameters,
+  UserOperationContext,
+} from "./types.js";
+
+export async function estimateUserOperationGas<
+  TTransport extends Transport = Transport,
+  TChain extends Chain | undefined = Chain | undefined,
+  TAccount extends SmartContractAccount | undefined =
+    | SmartContractAccount
+    | undefined,
+  TContext extends UserOperationContext | undefined =
+    | UserOperationContext
+    | undefined,
+  TEntryPointVersion extends GetEntryPointFromAccount<TAccount> = GetEntryPointFromAccount<TAccount>
+>(
+  client: Client<TTransport, TChain, TAccount>,
+  args: SendUserOperationParameters<TAccount, TContext>
+): Promise<UserOperationEstimateGasResponse<TEntryPointVersion>> {
+  const { account = client.account, overrides, uo } = args;
+  if (!account) {
+    throw new AccountNotFoundError();
+  }
+
+  if (!isBaseSmartAccountClient(client)) {
+    throw new IncompatibleClientError(
+      "BaseSmartAccountClient",
+      "estimateUserOperationGas",
+      client
+    );
+  }
+
+  const entryPoint = account.getEntryPoint();
+
+  const callData = Array.isArray(uo)
+    ? account.encodeBatchExecute(uo)
+    : typeof uo === "string"
+    ? uo
+    : account.encodeExecute(uo);
+
+  const signature = account.getDummySignature();
+
+  const nonce = account.getNonce(overrides?.nonceKey);
+
+  const struct =
+    entryPoint.version === "0.6.0"
+      ? ({
+          initCode: account.getInitCode(),
+          sender: account.address,
+          nonce,
+          callData,
+          signature,
+        } as Deferrable<UserOperationStruct<TEntryPointVersion>>)
+      : ({
+          factory: conditionalReturn(
+            account.isAccountDeployed().then((deployed) => !deployed),
+            account.getFactoryAddress()
```

Comment:
This looks slightly inefficient in that it will make the call to `getFactoryAddress` even in the case where it's already deployed, likewise with `getFactoryData` just below. It may be worth changing the signature here so the second argument to `conditionalReturn` is `() => account.getFactoryAddress()`.

Reply:
this is deferrable though, a promise right? so until at the end, these promises will not be actually executed/awaited?

Reply:
+1 to dphils comment the inputs to conditional return should be lambdas or at least the second parameter should be. They're promises, but they're instantiated immediately and will start running. calling `await` at somepoint just polls the completed state of the promise before unwrapping it

Reply:
makes sense. will udpate.